File: intermediate_data_structures/intermediate_ds/queue.py
from collections import deque
class Queue:
    def __init__(self): self._items = deque()
    def __len__(self): return len(self._items)
    def enqueue(self, value): self._items.append(value)
    def dequeue(self):
        if not self._items: raise IndexError("dequeue from empty queue")
        return self._items.popleft()
    def peek(self):
        if not self._items: raise IndexError("peek from empty queue")
        return self._items[0]

File: go-data-structures-journey/structures/queue.go
package structures

// Queue is a first-in, first-out collection.
type Queue[T any] struct{ items []T; head int }
func(q *Queue[T]) Enqueue(v T){q.items=append(q.items,v)}
func(q *Queue[T]) Dequeue()(T,bool){var z T;if q.head==len(q.items){return z,false};v:=q.items[q.head];q.head++;if q.head*2>=len(q.items){q.items=append([]T(nil),q.items[q.head:]...);q.head=0};return v,true}
func(q *Queue[T]) Len()int{return len(q.items)-q.head}
